feat(conductor): add raft config to allow no shutdown on remove

KyrinCode opened the pull request:
**Description**

This PR adds a new flag `--raft.no-shutdown-on-remove` to `op-conductor` that prevents Raft from shutting down when the node is removed from the cluster.

**Problem**

When a node is removed from the cluster via `conductor_removeServer`, HashiCorp Raft's default behavior (`ShutdownOnRemove=true`) causes the Raft instance to completely shut down. This means:
- The node cannot be re-added to the cluster without restarting the entire process
- This creates operational complexity when managing cluster membership dynamically
- In scenarios where nodes need to be temporarily removed and re-added (e.g., during maintenance or role changes), a full restart is required

**Solution**

Add a configurable flag that sets `ShutdownOnRemove=false` in Raft configuration. When enabled:
- The node transitions to follower state instead of shutting down when removed
- The node can be immediately re-added to the cluster via `conductor_addServerAsVoter` or `conductor_addServerAsNonvoter`
- No process restart is required

**Changes**

- Added `--raft.no-shutdown-on-remove` flag (default: `false`, backward compatible)
- Added `NoShutdownOnRemove` field to `RaftConsensusConfig`
- Configured Raft's `ShutdownOnRemove` based on the flag value

**Tests**

Manual testing was performed:
1. Started a 3-node conductor cluster with `--raft.no-shutdown-on-remove` enabled
2. Removed a node via `conductor_removeServer`
3. Verified the node transitioned to follower state (instead of Raft shutting down)
4. Re-added the node via `conductor_addServerAsVoter`
5. Verified the node successfully rejoined the cluster without restarting (by checking the `conductor_clusterMembership` of this node)

**Additional context**

This feature is opt-in and disabled by default to maintain backward compatibility with existing deployments. The default Raft behavior (shutdown on remove) is preserved unless explicitly configured otherwise.
Use cases for this flag:
- Dynamic cluster membership management without downtime
- Graceful role transitions (voter ↔ nonvoter)
- Easier disaster recovery scenarios

**Metadata**

- Related to cluster membership management and operational flexibility

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

yashvardhan-kukreja commented:
The PR looks good. I am just curious about trying to run it in a real devnet, and see how ok other routines of op-conductor act when the raft instance isn't shutdown despite it being called with `removeServer`. There might be some potential leaks or graceless errors.

I would be happy to approve after I test this out and see if there's no obvious unexpected behaviour with `--raft.no-shutdown-on-remove`

yashvardhan-kukreja commented:
Hi @KyrinCode , I tested this case using:
- a older image (the one built on your demote Voter PR)
- a new image built on this PR (without the command arg)
- a new image built on this PR (with the command arg)

In all of the cases, I got the same result which was being able to removeServer and addServerAsVoter without any hiccups / restarts (in fact, I was able to shift leadership to the newly added conductor)

So I am having a hard time understanding the need of this PR.

--------
**Removing**

```
❯ curl -X POST -H "Content-Type: application/json" --data \
    '{"jsonrpc":"2.0","method":"conductor_removeServer","params":["dev-infra-eris-0-opn-reth-f-sequencer-2", 0],"id":1}'  \
    https://eris-0-opn-geth-f-sequencer-0-op-conductor.primary.infra.dev.oplabs.cloud | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   153    0    39  100   114    119    348 --:--:-- --:--:-- --:--:--   469
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": null
}

❯ cast rpc -r=https://eris-0-opn-reth-f-sequencer-2-op-conductor.primary.infra.dev.oplabs.cloud conductor_clusterMembership | jq -r .
{
  "servers": [
    {
      "id": "dev-infra-eris-0-opn-geth-f-sequencer-0",
      "addr": "eris-0-opn-geth-f-sequencer-0-op-conductor-raft.primary.infra.dev.oplabs.cloud:50050",
      "suffrage": 0
    },
    {
      "id": "dev-infra-eris-0-opn-geth-f-sequencer-1",
      "addr": "eris-0-opn-geth-f-sequencer-1-op-conductor-raft.primary.infra.dev.oplabs.cloud:50050",
      "suffrage": 0
    }
  ],
  "version": 0
}
```

Meanwhile conductor logs
```
=2026-03-20T14:00:48+0000 lvl=info msg="sequencer is healthy"
t=2026-03-20T14:00:49+0000 lvl=info msg="sequencer is healthy"
2026-03-20T14:00:50.684Z [WARN]  raft: heartbeat timeout reached, not part of a stable configuration or a non-voter, not triggering a leader election
t=2026-03-20T14:00:50+0000 lvl=info msg="sequencer is healthy"
t=2026-03-20T14:00:51+0000 lvl=info msg="sequencer is healthy"
t=2026-03-20T14:00:52+0000 lvl=info msg="sequencer is healthy"
```

**Adding back**

```
❯ curl -X POST -H "Content-Type: application/json" --data \
    '{"jsonrpc":"2.0","method":"conductor_addServerAsVoter","params":["dev-infra-eris-0-opn-reth-f-sequencer-2", "eris-0-opn-reth-f-sequencer-2-op-conductor-raft.primary.infra.dev.oplabs.cloud:50050", 0],"id":1}'  \
    https://eris-0-opn-geth-f-sequencer-0-op-conductor.primary.infra.dev.oplabs.cloud | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   245    0    39  100   206    198   1046 --:--:-- --:--:-- --:--:--  1250
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": null
}

❯ cast rpc -r=https://eris-0-opn-reth-f-sequencer-2-op-conductor.primary.infra.dev.oplabs.cloud conductor_clusterMembership | jq -r .
{
  "servers": [
    {
      "id": "dev-infra-eris-0-opn-geth-f-sequencer-0",
      "addr": "eris-0-opn-geth-f-sequencer-0-op-conductor-raft.primary.infra.dev.oplabs.cloud:50050",
      "suffrage": 0
    },
    {
      "id": "dev-infra-eris-0-opn-geth-f-sequencer-1",
      "addr": "eris-0-opn-geth-f-sequencer-1-op-conductor-raft.primary.infra.dev.oplabs.cloud:50050",
      "suffrage": 0
    },
    {
      "id": "dev-infra-eris-0-opn-reth-f-sequencer-2",
      "addr": "eris-0-opn-reth-f-sequencer-2-op-conductor-raft.primary.infra.dev.oplabs.cloud:50050",
      "suffrage": 0
    }
  ],
  "version": 0
}
```

Meanwhile conductor logs

```
t=2026-03-20T14:01:05+0000 lvl=info msg="sequencer is healthy"
t=2026-03-20T14:01:06+0000 lvl=info msg="sequencer is healthy"
2026-03-20T14:01:07.606Z [WARN]  raft: failed to get previous log: previous-index=15633857 last-index=15633847 error="log not found"
t=2026-03-20T14:01:07+0000 lvl=info msg="sequencer is healthy"
t=2026-03-20T14:01:08+0000 lvl=info msg="sequencer is healthy"
```

yashvardhan-kukreja commented:
Hi @KyrinCode , humble reminder for the above comment.

KyrinCode commented:
> Hi @KyrinCode , humble reminder for the above comment.

Yes, man. Thanks for your new findings, I'm retesting this. 👍

KyrinCode commented:
Hi @yashvardhan-kukreja, thanks for the thorough testing! I investigated the raft source code and figured out why your test didn't reproduce the issue.

Root cause: ShutdownOnRemove only fires in leaderLoop()

Looking at hashicorp raft v1.7.3 ([raft.go#L838-L845](https://github.com/hashicorp/raft/blob/v1.7.3/raft.go#L838-L845)):

```
  // inside leaderLoop(), after committing a config change:
  if !hasVote(r.configurations.committed, r.localID) {
      stepDown = true
  }
  // ...
  if stepDown {
      if r.config().ShutdownOnRemove {
          r.logger.Info("removed ourself, shutting down")
          r.Shutdown()
      }
  }
```
This code only exists in the leader's main loop. When you remove a follower, the follower simply stops receiving heartbeats — it never evaluates ShutdownOnRemove. That's why removing sequencer-2 (a follower) from the leader worked fine in all your tests.

The actual trigger: leader removes itself

The issue only manifests when the leader is the target of removeServer. I reproduced it on a local 3-node devnet:

```
  # Leader removes itself
  LEADER_ID=$(cast rpc conductor_leaderWithID | jq -r .id)
  cast rpc conductor_removeServer "$LEADER_ID" 0
```

Without --raft.no-shutdown-on-remove:
raft: removed ourself, shutting down
  - Raft instance is fully shut down, process stays alive (RPC still responds)
  - New leader can call addServerAsVoter, but replication permanently fails: failed to contact: server-id=conductor-1
  - The node becomes a zombie — must restart to recover

With --raft.no-shutdown-on-remove:
raft: removed ourself, transitioning to follower
  - Raft transitions to follower, stays alive
  - addServerAsVoter works, node fully rejoins the cluster
  - Can even transfer leadership back to it — no restart needed

When does this happen in practice?

In automated failover scenarios — for instance, when the conductor's health check loop determines the leader is unhealthy and triggers transferLeader(). If the transfer fails and the system falls back to removing the unhealthy leader from the cluster, ShutdownOnRemove creates a zombie node.

yashvardhan-kukreja commented:
Awesome, I just tested this out, and it's working perfectly!

The most visible negative impact of shutdown-on-remove is the inability to transfer leadership to the removed-and-added conductor without a restart.

With your PR, it worked as expected.

<img width="862" height="671" alt="Screenshot 2026-03-25 at 16 31 01" src="https://github.com/user-attachments/assets/1ad351dc-52de-48e5-a2dd-b28f92eed263" />

yashvardhan-kukreja commented:
/ci authorize 4614c2f1f71f2fafa47e7dbef469183288d3ff5f

KyrinCode commented:
@yashvardhan-kukreja Cool.

Looking at this more carefully, I don't think there's a valid use case for default ShutdownOnRemove=true in op-conductor. A zombie node (raft dead, process alive) is never the desired outcome. I'd suggest we simply hardcode rc.ShutdownOnRemove = false in NewRaftConsensus() instead of adding an opt-in flag — this makes the fix unconditional and eliminates the risk of operators forgetting to set the flag.

```go
// op-conductor/consensus/raft.go, NewRaftConsensus()
  rc := raft.DefaultConfig()
  rc.SnapshotInterval = cfg.SnapshotInterval
  rc.TrailingLogs = cfg.TrailingLogs
  rc.SnapshotThreshold = cfg.SnapshotThreshold
  rc.HeartbeatTimeout = cfg.HeartbeatTimeout
  rc.LeaderLeaseTimeout = cfg.LeaderLeaseTimeout
  rc.LocalID = raft.ServerID(cfg.ServerID)
  // Prevent raft from shutting down when leader removes itself from cluster.
  // Without this, the node becomes a zombie (process alive, raft dead) that
  // cannot rejoin without a restart. With this, it transitions to follower.
  rc.ShutdownOnRemove = false
```

yashvardhan-kukreja commented:
Yep @KyrinCode , that works as well. Feel free to make that change.

KyrinCode committed:
fix(op-conductor): disable raft ShutdownOnRemove to prevent zombie nodes

Hardcode ShutdownOnRemove=false in raft config instead of using an opt-in flag.

When the leader removes itself from the cluster (e.g., during automated
failover), raft's default ShutdownOnRemove=true shuts down the raft instance
while the process stays alive. This creates a zombie node: RPC responds but
raft is dead, replication fails permanently, and the node cannot rejoin
without a container restart.

With ShutdownOnRemove=false, the node transitions to follower state and can
be cleanly re-added to the cluster. There is no valid use case for the
zombie behavior, so this is hardcoded rather than configurable.

yashvardhan-kukreja commented:
/ci authorize 0b29f9b9e07f697877cfc6025d401150034d4fe8
